File: leetcode/weekly-contest/20240324/100228-执行操作使数据元素之和大于等于K.cpp
#include <cmath>

namespace math {

class Solution
{
public:
  int minOperations(int k)
  {
    return ceil(2 * sqrt(k) - 2);
  }
};

}  // namespace math

namespace binary_search {

class Solution
{
public:
  int minOperations(int k)
  {
    if (k == 1)
      return 0;

    auto checkFn = [&](int times) {
      for (int i = 0; i <= times; ++i) {
        int elem = 1 + i;
        int sum  = elem;
        for (int j = times - i; j > 0; --j) {
          sum += elem;
        }
        if (sum >= k)
          return true;
      }
      return false;
    };

    int l = 1, r = k - 1;

    while (l < r) {
      int mid = (l + r) >> 1;
      // printf("l=%d, r=%d, mid=%d\n", l, r, mid);
      if (checkFn(mid)) {
        r = mid;
      } else {
        l = mid + 1;
      }
    }
    // printf("l=%d, r=%d\n", l, r);
    return l;
  }
};

}  // namespace binary_search

File: leetcode/weekly-contest/20240324/100228-执行操作使数据元素之和大于等于K.py
from cmath import sqrt
from math import ceil

"""
枚举
"""

class Solution:
    def minOperations(self, k: int) -> int:
        return min(m - 1 + (k - 1) // m for m in range(1, k + 1))

"""
二分
"""

class Solution:
    def minOperations(self, k: int) -> int:
        if k == 1:
            return 0

        def check(target: int) -> bool:
            for i in range(0, target + 1):
                val = 1 + i
                sum = val * (target - i + 1)
                if sum >= k: return True
            return False

        left, right = 1, k - 1
        while left < right:
            middle = (left + right) // 2
            if check(middle):
                right = middle
            else:
                left = middle + 1

        return left

"""
数学
"""
class Solution:
    def minOperations(self, k: int) -> int:
       return ceil(2*sqrt(k)-2)
 